wolfssl/src/ssl.rs

```diff
@@ -1100,6 +1111,95 @@ impl<IOCB: IOCallbacks> Session<IOCB> {
             e => unreachable!("{e:?}"),
         }
     }
+
+    /// Enable key logging for applications
+    #[cfg(feature = "debug")]
+    pub fn enable_key_logging(&mut self, keylog: Arc<dyn Tls13SecretCallbacks>) -> Result<()> {
+        self.keylog = Some(keylog.clone());
+
+        let keylog_c = Box::into_raw(Box::new(keylog));
+
+        // SAFETY: No documentation available for [`wolfSSL_KeepArrays`] [`wolfSSL_set_tls13_secret_cb`].
+        // But based on api implementation, it expects a valid pointer to `WOLFSSL`.
+        //
+        // This implementation is taken from following examples:
+        // https://github.com/wolfSSL/wolfssl-examples/blob/master/tls/client-tls13.c
+        // https://github.com/wolfSSL/wolfssl-examples/blob/master/tls/server-tls13.c
+        match unsafe {
+            let ssl = self.ssl.lock().as_ptr();
+            wolfssl_sys::wolfSSL_KeepArrays(ssl);
+            wolfssl_sys::wolfSSL_set_tls13_secret_cb(
+                ssl,
+                Some(Self::tls13_secret_cb),
+                keylog_c as *mut c_void,
+            )
+        } {
+            wolfssl_sys::WOLFSSL_SUCCESS => Ok(()),
+            wolfssl_sys::WOLFSSL_FATAL_ERROR => panic!("No SSL context"),
+            e => unreachable!("{e:?}"),
+        }
+    }
+
+    #[cfg(feature = "debug")]
+    unsafe extern "C" fn tls13_secret_cb(
+        ssl: *mut wolfssl_sys::WOLFSSL,
+        id: ::std::os::raw::c_int,
+        secret: *const ::std::os::raw::c_uchar,
+        secret_size: ::std::os::raw::c_int,
+        ctx: *mut ::std::os::raw::c_void,
+    ) -> ::std::os::raw::c_int {
+        debug_assert!(!ssl.is_null());
+        debug_assert!(!secret.is_null());
+        debug_assert!(!ctx.is_null());
+
+        let keylog = Box::from_raw(ctx as *mut Arc<dyn Tls13SecretCallbacks>);
```

Comment:
I'm concerned what happens when this `keylog` goes out of scope the first time -- I think it would free the pointer to `Arc` which it contains? (But see above for a theory on how to avoid the box entirely).